plugins/modules/openssh_cert.py

```diff
@@ -377,34 +394,83 @@ def _command_arguments(self, key_copy_path):
 
         return result
 
-    def _generate_diff(self):
-        before = self.original_data.to_dict() if self.original_data else {}
-        before.pop('nonce', None)
-        after = self.data.to_dict() if self.data else {}
-        after.pop('nonce', None)
+    def _compare_options(self):
+        try:
+            critical_options, extensions = parse_option_list(self.options)
+        except ValueError as e:
+            return self.module.fail_json(msg=to_native(e))
+
+        return all([
+            set(self.original_data.critical_options) == set(critical_options),
+            set(self.original_data.extensions) == set(extensions)
+        ])
+
+    def _compare_time_parameters(self):
+        try:
+            original_time_parameters = OpensshCertificateTimeParameters(
+                valid_from=self.original_data.valid_after,
+                valid_to=self.original_data.valid_before
+            )
+        except ValueError as e:
+            return self.module.fail_json(msg=to_native(e))
+
+        return all([
+            original_time_parameters == self.time_parameters,
+            original_time_parameters.within_range(self.valid_at)
+        ])
+
+    def _generate_temp_certificate(self):
+        key_copy = os.path.join(self.module.tmpdir, os.path.basename(self.public_key))
+
+        try:
+            self.module.preserved_copy(self.public_key, key_copy)
+        except OSError as e:
+            self.module.fail_json(msg="Unable to stage temporary key: %s" % to_native(e))
+        self.module.add_cleanup_file(key_copy)
 
-        return {'before': before, 'after': after}
+        self.module.run_command(self._command_arguments(key_copy), environ_update=dict(TZ="UTC"), check_rc=True)
+
+        temp_cert = os.path.splitext(key_copy)[0] + '-cert.pub'
+        self.module.add_cleanup_file(temp_cert)
+
+        return temp_cert
 
     def _get_cert_info(self):
         return self.module.run_command([self.ssh_keygen, '-Lf', self.path])[1]
 
+    def _get_key_fingerprint(self, path):
+        stdout = self.module.run_command([self.ssh_keygen, '-lf', path], check_rc=True)[1]
+        return stdout.split()[1]
+
     def _is_valid(self):
-        if self.original_data:
-            try:
-                original_time_parameters = OpensshCertificateTimeParameters(
-                    valid_from=self.original_data.valid_after,
-                    valid_to=self.original_data.valid_before
+        partial_result = all([
+            set(self.original_data.principals) == set(self.principals),
+            self.original_data.serial == self.serial_number if self.serial_number is not None else True,
+            self.original_data.type == self.type,
+            self._compare_time_parameters(),
+        ])
+
+        return partial_result if self.regenerate == 'partial_idempotence' else partial_result and all([
```

Comment:
I think this way it would be a bit easier to understand:
```suggestion
        if self.regenerate == 'partial_idempotence':
            return partial_result

        return partial_result and all([
```